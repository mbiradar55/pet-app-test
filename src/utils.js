const petList = ['Snow', 'Gold', 'Brown'];

export const getPetList = () => Promise.resolve([...petList]);

export const addPet = (name) => {
  const trimmedName = name.trim();
  if (trimmedName.length === 0) {
    return Promise.reject(new Error('Name is required.'));
  }
  if (petList.includes(name)) {
    return Promise.reject(new Error(`${name} is already in the list.`));
  }
  petList.push(name);
  return Promise.resolve([...petList]);
};

export const removePet = () => {
  if (petList.length === 0) {
    return Promise.reject(new Error('There are no pets to be removed.'));
  }
  petList.pop();
  return Promise.resolve([...petList]);
};