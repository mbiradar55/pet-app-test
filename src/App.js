import { useEffect, useState } from 'react';
import { addPet, getPetList, removePet } from './utils';

function App() {
  const [petName, setPetName] = useState('');
  const [petList, setPetList] = useState([]);

  // read a pet list during initialization
  useEffect(() => {
    getPetList()
      .then((response) => setPetList(response))
      .catch((error) => console.error(error.message));
  }, []);

  // add a pet by name
  const addOnePet = () => {
    addPet(petName)
      .then((response) => {
        setPetList(response);
        setPetName('');
      })
      .catch((error) => console.error(error.message));
  };

  // remove the last pet
  const removeLastPet = () => {
    removePet()
      .then((response) => setPetList(response))
      .catch((error) => console.error(error.message));
  };

  // get a message about pets
  const getMessage = () =>
    petList.length === 0
      ? 'You have no pets.'
      : `This is a list of your pets: ${petList.join(', ')}.`;

  return (
    <>
      <div>{getMessage()}</div>
      <input
        value={petName}
        onChange={(event) => setPetName(event.target.value)}
      />
      <button onClick={addOnePet}>Add One Pet</button>
      <button onClick={removeLastPet}>Remove Last Pet</button>
    </>
  );
}

export default App;