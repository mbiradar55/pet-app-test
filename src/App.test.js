import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import App from './App';
import * as utils from './utils';

describe('renders <App />', () => {
  beforeEach(jest.spyOn(console, 'error').mockImplementation(() => undefined));
  afterEach(() => jest.clearAllMocks());

  it('should pass mount snapshot test', async () => {
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should pet list test', async () => {
    utils.getPetList = jest.fn().mockResolvedValue(['Snow', 'Gold', 'Brown']);
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    expect(wrapper.text()).toContain(
      'This is a list of your pets: Snow, Gold, Brown.'
    );
    expect(utils.getPetList).toBeCalledTimes(1);
    utils.getPetList.mockRestore();
  });
  it('should show the failed text message', async () => {
    utils.getPetList = jest.fn().mockRejectedValue([]);
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    expect(wrapper.text()).toContain('You have no pets.');
    expect(utils.getPetList).toBeCalledTimes(1);
    utils.getPetList.mockRestore();
  });
  it('should be able to delete a pet', async () => {
    utils.getPetList = jest.fn().mockResolvedValue(['Snow', 'Gold', 'Brown']);
    utils.removePet = jest.fn().mockResolvedValue(['Snow', 'Gold']);
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    const deleteButton = wrapper.find('button').at(1);
    await act(async () => {
      deleteButton.simulate('click');
    });
    expect(wrapper.text()).toContain(
      'This is a list of your pets: Snow, Gold.'
    );
    expect(utils.getPetList).toBeCalledTimes(1);
    expect(utils.removePet).toBeCalledTimes(1);
  });
  it('should not be able to delete no pets', async () => {
    utils.getPetList = jest.fn().mockResolvedValue([]);
    utils.removePet = jest
      .fn()
      .mockRejectedValue(new Error('There are no pets to be removed.'));
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    const deleteButton = wrapper.find('button').at(1);
    await act(async () => {
      deleteButton.simulate('click');
    });
    expect(wrapper.text()).toContain('You have no pets.');
    expect(utils.getPetList).toBeCalledTimes(1);
    expect(utils.removePet).toBeCalledTimes(1);
  });
  it('should not be able to add empty pet name', async () => {
    utils.getPetList = jest.fn().mockResolvedValue(['Snow', 'Gold', 'Brown']);
    utils.addPet = jest.fn().mockRejectedValue(new Error('Name is required.'));
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    const addButton = wrapper.find('button').at(0);
    await act(async () => {
      addButton.simulate('click');
    });
    expect(wrapper.text()).toContain(
      'This is a list of your pets: Snow, Gold, Brown.'
    );
    expect(utils.getPetList).toBeCalledTimes(1);
    expect(utils.addPet).toBeCalledTimes(1);
  });
  it('should be able to add a pet with name', async () => {
    const initialList = ['Snow', 'Gold', 'Brown'];
    utils.getPetList = jest.fn().mockResolvedValue(initialList);
    utils.addPet = jest
      .fn()
      .mockImplementation((name) => Promise.resolve([...initialList, name]));
    let wrapper;
    await act(async () => {
      wrapper = mount(<App />);
    });
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: 'Yellow' } });
    const addButton = wrapper.find('button').at(0);
    await act(async () => {
      addButton.simulate('click');
    });
    expect(wrapper.text()).toContain(
      'This is a list of your pets: Snow, Gold, Brown, Yellow.'
    );
    expect(utils.getPetList).toBeCalledTimes(1);
    expect(utils.addPet).toBeCalledTimes(1);
  });
});
