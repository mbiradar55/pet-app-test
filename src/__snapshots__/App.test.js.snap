// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <App /> should pass mount snapshot test 1`] = `ReactWrapper {}`;
